src/cabin/cabin.service.ts

```diff
@@ -1,14 +1,115 @@
 import { Injectable } from '@nestjs/common';
+import { VisbookApi } from 'src/visbook/visbook.api';
 import { CabinDetails, CabinApi } from './cabin.api';
+import { dateIsValid, getVisbookId } from './cabin.utils';
 
 @Injectable()
 export class CabinService {
   private readonly apiClient = new CabinApi();
+  private readonly visbookApiClient = new VisbookApi();
+
   async getRandomCabin(): Promise<CabinDetails> {
     const cabins = await this.apiClient.getCabins();
     const cabin = cabins[Math.floor(Math.random() * cabins.length)];
     const { id } = cabin.node;
     const cabinDetails = await this.apiClient.getCabinDetails(id);
     return cabinDetails;
   }
+
+  async isCabinAvailable(
+    cabin: CabinDetails,
+    checkIn: string,
+    checkOut: string,
+  ): Promise<boolean> {
+    if (!dateIsValid(checkIn) || !dateIsValid(checkOut)) {
+      console.log('invalid date(s)');
+      return false;
+    }
+
+    if (cabin.bookingUrl === null) {
+      console.log('no booking url');
+      return false;
+    }
+
+    if (!cabin.bookingUrl.includes('reservations')) {
+      console.log('not a reservations url');
+      return false;
+    }
+
+    const cabinVisbookId = getVisbookId(cabin.bookingUrl);
+
+    if (cabinVisbookId === 0) return false;
+
+    const cabinIsAvailable = await this.visbookApiClient.isCabinAvailable(
+      cabinVisbookId,
+      checkIn,
+      checkOut,
+    );
+
+    return cabinIsAvailable;
+  }
+
+  async getRandomAvailableCabin(
+    checkIn: string,
+    checkOut: string,
+  ): Promise<CabinDetails | undefined> {
+    if (!dateIsValid(checkIn) || !dateIsValid(checkOut)) {
+      return undefined;
+    }
+
+    const hasExceededTimeLimit = (now: number, timeLimit: number): boolean => {
+      const elapsedTime = now - startTime;
+      if (elapsedTime >= timeLimit) {
+        return true;
+      }
+      return false;
+    };
+
+    const startTime = Date.now();
+    let cabin = await this.getRandomCabin();
+
+    while ((await this.isCabinAvailable(cabin, checkIn, checkOut)) === false) {
+      cabin = await this.getRandomCabin();
+      if (hasExceededTimeLimit(Date.now(), 60000)) {
+        /* TODO
+        - How long should response time (time limit) be? 1min? 5min?
+        - Do we have a limit on how many api calls we can make in a given time?
+        - Return undefined or specific message aka "could not find within time limit?"
+        */
```

Comment:
How long does this take from your testing? I think we can start with something for now and then go from there.

This sounds like another good reason to avoid using their API directly for this information.

Reply:
Right now it's not so busy in the cabins, so it usually doesn't take more than 1 to 3 attempts to find an available one. Result came after 5 to 20s, but I think a minute is fair to be prepared for busier times.
Note that this feature in the current implementation also depends on a call to the cabin api (to get cabin details such as booking url). Once we have basic cabin information stored in our database, we won't need to call two apis anymore.

Reply:
Cool. Good to know